### PR TITLE
Replace `yapf` with `black`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

The addition of `--skip-magic-trailing-comma` to `black` sparked this discussion again.

cc: people interested about this in the past: @akihironitta @Borda @ananthsub @awaelchli @edgarriba 

Pros for black:
- Faster (https://github.com/kornia/kornia/pull/1051#issuecomment-851550383)
- More active project
- More popular project (21.2k stars vs 11.7k)
- More consistent, tighter constraints
- Also formats docstrings
- Removes `\` to split lines

Cons:
- Not as configurable

## Vote

- Yay (👍) or nay (👎)?
- Default `"` (🚀) or `'` (🎉) for strings?

**If 👍, this won't be merged until after the last bug-fix release in the `v1.3.x` line to avoid conflicts.**

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [n/a] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Comment by awaelchli** [comment]
@carmocca configurability of code style can be named as a pro for yapf but not as a con for black because it is designed that way.

**Comment by carmocca** [comment]
These are the stats for the diff size for each string option:

`'`: 355 files changed, 5310 insertions(+), 7858 deletions(-)
`"`: 335 files changed, 5527 insertions(+), 8075 deletions(-)

**Comment by Borda** [comment]
@carmocca mind sho how the codebase would look like?

File: benchmarks/test_sharded_parity.py [hunk 1]
```diff
@@ -144,27 +142,15 @@ def plugin_parity_test(
...
     use_cuda = gpus > 0
 
-    trainer = Trainer(
-        fast_dev_run=True,
-        max_epochs=1,
-        gpus=gpus,
-        precision=precision,
-        accelerator='ddp_spawn',
-    )
+    trainer = Trainer(fast_dev_run=True, max_epochs=1, gpus=gpus, precision=precision, accelerator="ddp_spawn")
```
Comment (Borda) [inline benchmarks/test_sharded_parity.py:145]: I would prefer to keep this as multiline :confused:

File: pl_examples/basic_examples/conv_sequential_example.py [hunk 2]
```diff
@@ -173,17 +170,16 @@ def configure_optimizers(self):
...
-        cifar10_normalization(),
-    ])
+    train_transforms = torchvision.transforms.Compose(
+        [
+            torchvision.transforms.RandomCrop(32, padding=4),
+            torchvision.transforms.RandomHorizontalFlip(),
+            torchvision.transforms.ToTensor(),
+            cifar10_normalization(),
+        ]
+    )
```
Comment (Borda) [inline pl_examples/basic_examples/conv_sequential_example.py:180]: the past collapsing was cuter :rabbit:

File: benchmarks/test_sharded_parity.py (hunk 1 above)
Comment (carmocca) [reply benchmarks/test_sharded_parity.py:145]: > I would prefer to keep this as multiline 😕

why? It's one line because it fits in the 120 max length.

So it's consistent.

> would need the magic trailing comma right?

Yes but we are using `--skip-magic-trailing-comma` so since it can be in a single line, it would get removed.

Comment (Borda) [reply benchmarks/test_sharded_parity.py:145]: > why? It's one line because it fits in the 120 max length.

yes, but is about readability, have the trainer formatted always the same way, so for example as multiline so you just simply look and see, also any argument change is much easier to recognise, then mentally parsing full line :D

Comment (awaelchli) [reply benchmarks/test_sharded_parity.py:145]: I vote for having the magic comma with this behavior: 
https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#the-magic-trailing-comma

File: pl_examples/basic_examples/conv_sequential_example.py (hunk 2 above)
Comment (carmocca) [reply pl_examples/basic_examples/conv_sequential_example.py:180]: I believe the reason for this is that so if you pass more arguments to `Compose`, your diff will impact less lines:

```python
result = a_long_function_name([
    "the red fox jumped",
    "over the fence",
    "foo bar bar",
])
```

later...
```python
result = a_long_function_name([
    "the red fox jumped",
    "over the fence",
    "foo bar foobar",
], "another_argument", "maybe_one_more")
```

File: benchmarks/test_sharded_parity.py (hunk 1 above)
Comment (carmocca) [reply benchmarks/test_sharded_parity.py:145]: The issue with the magic comma is that you have to be mindful about where you add it, as the formatted does not know whether you manually added it because you want a specific line to be expanded or because it was there accidentally.

For example, take this input snippet with different alternatives:

```python
if isinstance(input, (list, tuple)):
    pass

if isinstance(input, (list, tuple,),):
    pass

if isinstance(
    input,
    (
        list,
        tuple,
    ),
):
    pass

if isinstance(
    input,
    (
        list,
        tuple
    )
):
    pass
```

`black --skip-magic-trailing-comma file.py`

```python
if isinstance(input, (list, tuple)):
    pass

if isinstance(input, (list, tuple)):
    pass

if isinstance(input, (list, tuple)):
    pass

if isinstance(input, (list, tuple)):
    pass
```

`black file.py`

```python
if isinstance(input, (list, tuple)):
    pass

if isinstance(
    input,
    (
        list,
        tuple,
    ),
):
    pass

if isinstance(
    input,
    (
        list,
        tuple,
    ),
):
    pass

if isinstance(input, (list, tuple)):
    pass
```

Considering the whole expression can be fit into one line (because it's length is <120), I personally don't like that it stays expanded because a trailing comma was there.

Comment (carmocca) [reply benchmarks/test_sharded_parity.py:145]: > have the trainer formatted always the same way, so for example as multiline so you just simply look and see, also any argument change is much easier to recognise, then mentally parsing full line :D

The formatter checks if the line can fit under the line length limit, and formats it that way.

I would find it weird (and inconsistent) that we would break that rule one class in particular (`Trainer`)

File: pl_examples/basic_examples/conv_sequential_example.py (hunk 2 above)
Comment (ananthsub) [reply pl_examples/basic_examples/conv_sequential_example.py:180]: +1 to @carmocca's comment. this becomes very useful for preserving git blame

**Comment by ananthsub** [comment]
+1000000 strongly in favor of using black as a standard formatter :)

File: benchmarks/test_sharded_parity.py (hunk 1 above)
Comment (tchaton) [reply benchmarks/test_sharded_parity.py:145]: Definitely a hard problem :) 

I definitely prefer `black --skip-magic-trailing-comma file.py` in the above context but agree with Jirka on readability of expending Trainer arguments onto multiple lines in the tests.

I guess we can't have everything ;)

Comment (Borda) [reply benchmarks/test_sharded_parity.py:145]: we can fake it by repeating default args to have the line sufficiently long to get it split to multiple :D

**Review by SeanNaren — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM ! @carmocca some conflicts to resolve.

**Review by awaelchli — APPROVED** [review]
suggestion: make this the very last commit before 1.4 :)

**Review by Borda — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Comment by Borda** [comment]
@carmocca note that this big PR will be always in conflict whenever we merge to master, also after this merge most open PRs will be affected... :rabbit:

**Comment by carmocca** [comment]
> note that this big PR will be always in conflict whenever we merge to master

This is not a big deal, we just have to delete `pre-commit-bot`'s commit and let it re-do it to resolve conflicts. And obviously have a window of time where the CI for this PR runs before any other one is merged.

> also after this merge most open PRs will be affected... 🐰

Yes, but regardless, the open PRs can use the "keep ours" merge strategy and `pre-commit-bot` will fix it afterwards

**Comment by Borda** [comment]
just rebased and fixed conflicts... :]

**Review by ananthsub — APPROVED** [review]
Exciting!

**Comment by awaelchli** [comment]
@carmocca @Borda now that we use black, are you ok with removing pep8speaks? Whatever pep8 says we possibly cannot influence it due to the automatic formatting, so I would like to suggest removing these messages.

**Comment by carmocca** [comment]
> @carmocca @Borda now that we use black, are you ok with removing pep8speaks? Whatever pep8 says we possibly cannot influence it due to the automatic formatting, so I would like to suggest removing these messages. 

https://github.com/PyTorchLightning/pytorch-lightning/pull/8392#issuecomment-892609542

**Comment by Borda** [comment]
we l need to uninstall it from all repos I guess, will sync with @lexierule
